fuel_gauge: Add battery cutoff support

aaronemassey opened the pull request:
Many fuel gauge ICs offer a battery cutoff/shipping mode functionality that cutoff charge from the battery. This is often useful for preserving battery charge on devices while in storage.

Add battery cutoff support to the fuel gauge API with a generic default SBS driver showing an example of support in tests.

Please see  https://github.com/zephyrproject-rtos/zephyr/issues/61434 for more information on motivation and a graphic to gain some quick context on this PR.

keith-zephyr commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.h:
```diff
@@ -47,8 +47,33 @@
+ * gauge. For the case where it's a singular value that must only be written to the fuel gauge only
+ * once, retransmitting the duplicate write has no significant negative consequences.
+ *
+ * Why not devicetree: Finding the maximum length of all the battery cutoff payloads in a devicetree
+ * at compile-time would require labyrinthine amount of macro-batics.
+ *
```
I think this is a fine trade off for this use case.

As an alternative, you could allocate an array of `cutoff_payload` for each fuel gauge based on a devictree property length and then link to the array from the driver's static `config`.   This [example](https://github.com/zephyrproject-rtos/zephyr/blob/b42331476d3d88a16a3e82c9ac83ea7864fdfc61/drivers/gpio/gpio_nct38xx_alert.c#L132) uses the `DT_INST_FOREACH_PROP_ELEM` to allocate the array.  However, this approach would likely consume a little more memory than what you've done here.

aaronemassey replied:
Ack. Will keep current solution for now. Not resolving comment so other reviewers have visibility on this.

keith-zephyr commented on dts/bindings/fuel-gauge/battery_cutoff.yaml:
```diff
@@ -0,0 +1,28 @@
+  battery cutoff support. See the default fuel gauge SBS Gauge compatible as an example.
+
+properties:
+  battery_cutoff_support:
```
Devicetree [property names](https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#naming-conventions) should not use underscores.
```suggestion
  battery-cutoff-support:
```

keith-zephyr commented on dts/bindings/fuel-gauge/battery_cutoff.yaml:
```diff
@@ -0,0 +1,28 @@
+    description: |
+      Helper prop that indicates whether this device can cutoff the battery, this is also often
+      referred to as ship or sleep mode. Any non-zero value means battery cutoff is supported.
+    type: int
```
Why not bool type?

keith-zephyr commented on dts/bindings/fuel-gauge/battery_cutoff.yaml:
```diff
@@ -0,0 +1,28 @@
+    type: int
+  battery_cutoff_payload:
+    description: |
+       Payload to write to cutoff battery. This is often a sequence of two duplicate values.
```
```suggestion
       Payload to write to cutoff battery register. This must be an array of 2 integers.  
       If your battery only requires a single payload, duplicate the value.  This will 
       cause the driver to issue the cutoff twice.
```

keith-zephyr commented on dts/bindings/fuel-gauge/sbs,default-sbs-gauge.yaml:
```diff
@@ -0,0 +1,23 @@
+  actuated by the SBS driver for SBS compliant fuel gauge ICs.
+
+properties:
+  battery_cutoff_support:
+    default: 0
```
If you switch this to a bool, then you don't need to set the default value in the yaml schema.

keith-zephyr commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:
```diff
@@ -307,17 +331,39 @@ static const struct fuel_gauge_driver_api sbs_gauge_driver_api = {
+	.cutoff_reg_addr = DT_PROP_OR(DT_DRV_INST(index), battery_cutoff_reg_addr, 0),             \
+	.cutoff_payload = DT_PROP_OR(DT_DRV_INST(index), battery_cutoff_payload, {}),
+#else
+#define SBS_GAUGE_CONFIG_INIT(index) \
```
Unnecessary line continuation.
```suggestion
#define SBS_GAUGE_CONFIG_INIT(index)
```

keith-zephyr commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:
```diff
@@ -307,17 +331,38 @@ static const struct fuel_gauge_driver_api sbs_gauge_driver_api = {
-/* FIXME: fix init priority */
+#ifdef SBS_GAUGE_CUTOFF_SUPPORT
+#define SBS_GAUGE_CONFIG_INIT(index)                                                               \
+	.cutoff_support = DT_PROP_OR(DT_DRV_INST(index), battery_cutoff_support, false),           \
```
For boolean types, I believe `DT_NODE_HAS_PROP` is always true.  So I think this can be simplified:
```suggestion
	.cutoff_support = DT_INST_PROP(index), battery_cutoff_support),           \
```

keith-zephyr commented on dts/bindings/fuel-gauge/battery_cutoff.yaml:
```diff
@@ -0,0 +1,30 @@
+    type: int
+  battery-cutoff-payload:
+    description: |
+       Payload to write to cutoff battery register. This must be array of 2 integers. If your
+       battery only requires a single payload, duplicate the value.  This will cause the driver to
+       issue the cutoff twice.
```
Looking closer at your macros, you actually support a payload length of 1 or 2.  Change this help to note you can specify a maximum of 2 integers and drop the text about setting a duplicate value.

aaronemassey replied:
Yeah, though I was fine with being restrictive at first and just relaxing that as needed.

keith-zephyr commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:
```diff
@@ -307,17 +331,38 @@ static const struct fuel_gauge_driver_api sbs_gauge_driver_api = {
+#define SBS_GAUGE_VALIDATE_CUTOFF_PAYLOAD_SIZE(index)                                              \
+	(true _SBS_GAUGE_VALIDATE_CUTOFF_PAYLOAD_SIZE(index))
+
+BUILD_ASSERT(DT_INST_FOREACH_STATUS_OKAY(SBS_GAUGE_VALIDATE_CUTOFF_PAYLOAD_SIZE));
```
This can be simplified as:
```
#define CUTOFF_PAYLOAD_SIZE_ASSERT(inst) \
    BUILD_ASSERT(DT_INST_PROP_OR(inst, battery_cutoff_payload, 0) <=   \
        SBS_GAUGE_CUTOFF_PAYLOAD_MAX_SIZE)
DT_INST_FOREACH_STATUS_OKAY(CUTOFF_PAYLOAD_SIZE_ASSERT)
```

keith-zephyr commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:
```diff
@@ -307,17 +331,38 @@ static const struct fuel_gauge_driver_api sbs_gauge_driver_api = {
+#ifdef SBS_GAUGE_CUTOFF_SUPPORT
+#define SBS_GAUGE_CONFIG_INIT(index)                                                               \
+	.cutoff_support = DT_PROP_OR(DT_DRV_INST(index), battery_cutoff_support, false),           \
+	.cutoff_reg_addr = DT_PROP_OR(DT_DRV_INST(index), battery_cutoff_reg_addr, 0),             \
+	.cutoff_payload = DT_PROP_OR(DT_DRV_INST(index), battery_cutoff_payload, {}),
```
```suggestion
	.cutoff_reg_addr = DT_INST_PROP_OR(battery_cutoff_reg_addr, 0),             \
	.cutoff_payload = DT_INST_PROP_OR(battery_cutoff_payload, {}),
```

keith-zephyr commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.h:
```diff
@@ -47,8 +47,33 @@
 
+/*
+ * Nearly all cutoff payloads are actually a singular value that must be written twice to the fuel
+ * gauge. For the case where it's a singular value that must only be written to the fuel gauge only
+ * once, retransmitting the duplicate write has no significant negative consequences.
```
Delete reference to retransmitting duplicate value, since your code doesn't require that.

keith-zephyr commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:
```diff
@@ -173,6 +177,27 @@ static int sbs_gauge_get_prop(const struct device *dev, struct fuel_gauge_get_pr
+		return -ENOTSUP;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(cfg->cutoff_payload); i++) {
```
```suggestion
	for (int i = 0; i < cfg->cutoff_payload_size; i++) {
```

aaronemassey replied:
Actually `cutoff_payload_size` only exists in the emulator as a helper to prevent usages of `ARRAY_SIZE()`.

But  I realize that's a bit confusing so I'll remove the emulator's cutoff_payload_size in favor of `ARRAY_SIZE()`.

keith-zephyr replied:
The driver should capture a `cutoff_payload_size`.  `ARRAY_SIZE(cfg->cutoff_payload)` will always be equal to `SBS_GAUGE_CUTOFF_PAYLOAD_MAX_SIZE`.

This code will write a 0 to the cutoff register if the user only provides a single payload value.

aaronemassey replied:
Oh jeez. I should've jumped off earlier looking back at that this morning. Thanks!

keith-zephyr commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:
```diff
@@ -173,6 +177,27 @@ static int sbs_gauge_get_prop(const struct device *dev, struct fuel_gauge_get_pr
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(cfg->cutoff_payload); i++) {
+		rc = sbs_cmd_reg_write(dev, cfg->cutoff_reg_addr, cfg->cutoff_payload[0]);
```
Looks like you need a test case where the payload value is not duplicated.

```suggestion
		rc = sbs_cmd_reg_write(dev, cfg->cutoff_reg_addr, cfg->cutoff_payload[i]);
```

sjg20 commented on drivers/fuel_gauge/sbs_gauge/emul_sbs_gauge.c:
```diff
@@ -46,8 +57,48 @@ struct sbs_gauge_emul_data {
 	uint16_t addr;
+	bool cutoff_support;
+	uint32_t cutoff_reg_addr;
+	int cutoff_payload[SBS_GAUGE_CUTOFF_PAYLOAD_MAX_SIZE];
```
Can you add comments for these?

How about s/cutoff_reg_addr/cutoff_reg/ ?

sjg20 commented on drivers/fuel_gauge/sbs_gauge/emul_sbs_gauge.c:
```diff
@@ -46,8 +57,48 @@ struct sbs_gauge_emul_data {
+		 * We thoroughly check bounds elsewhere, so we can be confident we're not indexing
+		 * past the end of the array.
+		 */
+		uint16_t target_payload_elem_val =
```
This is u16 but above it uses int

sjg20 commented on dts/bindings/fuel-gauge/battery_cutoff.yaml:
```diff
@@ -0,0 +1,30 @@
+
+description: |
+
+  Properties for fuel-gauges that may control battery cutoff, this is common in SBS compliant or
```
SBS-compliant

sjg20 commented on dts/bindings/fuel-gauge/battery_cutoff.yaml:
```diff
@@ -0,0 +1,30 @@
+description: |
+
+  Properties for fuel-gauges that may control battery cutoff, this is common in SBS compliant or
+  similarly smart battery fuel gauges.
```
Do we need the word 'battery' ?

aaronemassey replied:
Would prefer to differentiate from other future forms of `cutoff` such as with charging.

Alternatively, I could rename all of this to ship_mode. I prefer the battery cutoff nomenclature because it's more self-defining.

sjg20 commented on dts/bindings/fuel-gauge/battery_cutoff.yaml:
```diff
@@ -0,0 +1,30 @@
+  similarly smart battery fuel gauges.
+
+  Note: These properties are to be used with meaningful defaults in fuel gauge ICs that implement
+  battery cutoff support. See the default fuel gauge SBS Gauge compatible as an example.
```
Drop the word 'support' ?

What does 'default fuel gauge SBS Gauge compatible' mean? It might be missing hyphens, but I'm not sure

aaronemassey replied:
I think (hope) I made it more clear in the following push.

sjg20 commented on dts/bindings/fuel-gauge/battery_cutoff.yaml:
```diff
@@ -0,0 +1,30 @@
+properties:
+  battery-cutoff-support:
+    description: |
+      Helper prop that indicates whether this device can cutoff the battery, this is also often
```
battery. This

(or use a semicolon?)

sjg20 commented on dts/bindings/fuel-gauge/sbs,default-sbs-gauge.yaml:
```diff
@@ -0,0 +1,22 @@
+include: ["sbs,sbs-gauge-new-api.yaml", "battery_cutoff.yaml"]
+
+description: |
+  Default generic smart battery fuel gauge driver. Includes support for battery cutoff if enabled.
```
I think nearly everywhere you use this it should be 'battery-cutoff'

sjg20 replied:
Perhaps you should just say 'cutoff' an then mention in one place that it means cutting off the battery?

aaronemassey replied:
I'd like to keep it to battery cutoff to be more explicit and differentiating from charge cutoff.

sjg20 commented on include/zephyr/drivers/emul_fuel_gauge.h:
```diff
@@ -34,6 +34,7 @@ extern "C" {
  */
 __subsystem struct fuel_gauge_emul_driver_api {
 	int (*set_battery_charging)(const struct emul *emul, uint32_t uV, int uA);
+	int (*is_battery_cutoff)(const struct emul *emul, bool *cutoff);
```
Do these not get comments?

aaronemassey replied:
Typically no. The client does not typically interact with the API struct directly but through functions defined in the same header.

sjg20 commented on include/zephyr/drivers/emul_fuel_gauge.h:
```diff
@@ -66,6 +67,26 @@ static inline int emul_fuel_gauge_set_battery_charging(const struct emul *target
 }
 
+/**
+ * @brief Check if the underlying emulator has battery cutoff
```
I believe this is checking whether the battery is actually cut off, not whether the emulator has the 'cutoff' feature.

Perhaps it would be easier if you just said 'Check if the battery is cut off' ?

sjg20 reviewed (dismissed):
LGTM with some doc nits

fabiobaltieri commented:
The regulator APIs have a call for shipping mode:

https://github.com/zephyrproject-rtos/zephyr/blob/03905f7e55d24788e0a38d1641dcadf529fce85d/include/zephyr/drivers/regulator.h#L67

No arguments or anything fancy, just a place to put the code, would it make sense to use that one instead?

aaronemassey commented:
> The regulator APIs have a call for shipping mode:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/03905f7e55d24788e0a38d1641dcadf529fce85d/include/zephyr/drivers/regulator.h#L67
> 
> No arguments or anything fancy, just a place to put the code, would it make sense to use that one instead?

So doing this would require creating multiple device instances and them having them coordinate and be cohesive. This seems like more trouble than it'd be worth over duplicating this in the fuel gauge API given the current constraints on Zephyr's device driver model.

I'd appreciate @keith-zephyr adding a word here.

fabiobaltieri commented:
> A battery fuel gauge is not a regulator, so I don't see adding 'cutoff' to batteries as redundant with regulators. It just so happens both device types provide a similar function. Trying to shoehorn this into the regulator parent would also mean the fuel gauge would need to stub out `dvs_state_set`.

Ah you could leave that unset and let the API stub check it. Was thinking you could do a child device just for that function but yeah that does seem a bit overkill. Thing is that this functionality can go to few different places, for instance https://www.ti.com/product/BQ25180 is a lipo charger and also has a ship mode, so we'd end up with an api for that in regulator, one in fuel gauge (which probably applies only for SBS batteries?) and one for chargers.

keith-zephyr commented on dts/bindings/fuel-gauge/battery-cutoff.yaml:
```diff
@@ -0,0 +1,29 @@
+  battery-cutoff-support:
+    description: |
+      Helper prop that indicates whether this device can cutoff the battery; this is also often
+      referred to as ship or sleep mode. Any non-zero value means battery cutoff is supported.
```
This is now a boolean property, so the "non-zero" comment no longer applies.
```suggestion
      referred to as ship or sleep mode.
```

keith-zephyr commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:
```diff
@@ -307,17 +331,31 @@ static const struct fuel_gauge_driver_api sbs_gauge_driver_api = {
 DT_INST_FOREACH_STATUS_OKAY(SBS_GAUGE_INIT)
+
+#define CUTOFF_PAYLOAD_SIZE_ASSERT(inst)                                                           \
+	BUILD_ASSERT(DT_PROP_LEN_OR(DT_DRV_INST(inst), battery_cutoff_payload, 0) <=               \
```
```suggestion
	BUILD_ASSERT(DT_INST_PROP_LEN_OR(inst, battery_cutoff_payload, 0) <=               \
```

aaronemassey replied:
Done.

fabiobaltieri commented:
> `twister-build (1)` build failure is unrelated:
> 
> ```
> devicetree error: /__w/zephyr/zephyr/boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:105 (column 1): parse error: undefined node label 'can1'
> ```

Should be fixed, reran CI for you.

aaronemassey committed:
fuel_gauge: Add battery cutoff support

Many fuel gauge ICs offer a battery cutoff/shipping mode functionality that
cutoff charge from the battery. This is often useful for preserving battery
charge on devices while in storage.

Add battery cutoff support to the fuel gauge API with a generic default SBS
driver showing an example of support in tests.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey commented:
Whoops accidentally combined with a prior commit and pushed. Fixed in https://github.com/zephyrproject-rtos/zephyr/compare/1322276e2bf9f6985a9ad58e5ad1bd81e95db701..3e4b314c7c183762e38a9f4a3ef0c670a9c1ae1c

fabiobaltieri commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.h:
```diff
@@ -47,8 +47,38 @@
+ */
+#define SBS_GAUGE_CUTOFF_PAYLOAD_MAX_SIZE 2
+
+#define _SBS_GAUGE_INST_CUTOFF_SUPPORTED(inst)                                                     \
+	DT_PROP_OR(DT_DRV_INST(inst), battery_cutoff_support, false)
+#define SBS_GAUGE_CUTOFF_SUPPORT()                                                                 \
+	(false || DT_INST_FOREACH_STATUS_OKAY(_SBS_GAUGE_INST_CUTOFF_SUPPORTED))
```
Interesting approach :-)

How about instead of this you split the cutoff fields in a separate config struct, only instantiate it if the specific instance has the property and then leave the handler there all the time (it only takes few bytes of flash) and return an error instantly if this structure is not set? (meaning the battery does not support it)

Or I guess you could rip off the multi instance support if we don't need it, though that'd be sad, I really miss my X250 :_(

fabiobaltieri replied:
Oh you already have `cutoff_support`, so yeah you could replace that with a struct pointer, if it's null you don't have cutoff support, if it's set you do and the data is there.

aaronemassey replied:
Sgtm, having the extra struct is easier to manage, gives grouping to the associated properties.

aaronemassey replied:
@fabiobaltieri But allocating a different struct makes no change to needing `SBS_GAUGE_CUTOFF_SUPPORT()` unless I'm missing something?

You're talking about just collecting constant parameters as a separate struct, right?

fabiobaltieri replied:
You should be able to drop `SBS_GAUGE_CUTOFF_SUPPORT()` entirely if you only allocate the extra structure if `battery_cutoff_support` is there, then `sbs_gauge_do_battery_cutoff` is always compiled in but validates that the structure is not there and returns `-ENOSUP` in that case.

aaronemassey replied:
Oh that's certainly cleaner. I'm totally happy to do that if we don't care about `sbs_gauge_do_battery_cutoff` always being compiled in and nominally increasing the code size.

However, I think it's more broadly suitable if we keep it compiled out. Battery cutoff is *NOT* standard SBS behavior so many folks might be willing to use this driver without wanting the battery cutoff feature.

I'll do that then. We can always revert it back to using the helper macro if requested :)

fabiobaltieri replied:
You can always have a Kconfig option to compile that out if one really wants to save those few bytes.

aaronemassey commented:
Last push was rebase only.

fabiobaltieri commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.h:
```diff
@@ -47,8 +47,35 @@
+
 struct sbs_gauge_config {
 	struct i2c_dt_spec i2c;
+	struct sbs_gauge_battery_cuttoff_config cutoff_cfg;
```
if you make this a pointer instead, then you can allocate it only if the prop it set and leave it NULL if not. Then drop `cutoff_support` and instead just do `if (cfg->cutoff_cfg = NULL) { return; }`

aaronemassey replied:
Yeah looking back at this after I'm not sure why I didn't realize this is what you originally meant. Fridays are fridays.
